### ISSUE TITLE
Optimizing Webpack Performance. Initial: 36s Incremental: 22s ...

### ISSUE THREAD

**Issue by maximerassi** [webpack/webpack#1325]
We have just got our fist working version of the app ported to webpack. The performance with everything back is now really poor and we have started trying to optimize it. Below are some things we have tried and documented below. We would greatly appreciate some feedback.

  * What else could we do?
  *  Is our design of relying on one entry point and using require.ensure excessively the reason for such slow build times? Or can we somehow get the compilation speed reasonable under such design.

## Previous Setup
We were previously using **gulp** which bundled everything into one file and was quite fast. With the use of plugins such as **gulp-remember** and **gulp-cached** we essentially had incremental builds of under 1 seconds, however that workflow had its own problems of its own. We are happy with the require.ensure possibilities with webpack and hopefully with your help we can get close to that or anywhere reasonable.

## Current Setup
At the current time we are getting an **Incremental Build Time** of 22.09 seconds with **Initial Build Time** of 36.11 seconds. Note we are heavily heavily using **require.ensure**. This feature made perfect sense for us as many of the modules (modes) are only available to the user in certain conditions so no reason to have it download by default.
This is currently generating **95** js bundle files creates. We have only one entry point 'initialize' that initializes our entry app and we make use of our framework's router to then require.ensure other pieces of the app based on the URL.

![2015-08-04_12-50-57](https://cloud.githubusercontent.com/assets/1452653/9066884/197dc2e6-3aa8-11e5-964b-bd7ca68879a3.jpg)

### Machine (Laptop):
  * i7 4800MQ 2.70 GHz
  * 16 GB RAM
  * SSD Disk Drive

### Machine (Laptop):
  * npm 2.10.1
  * webpack 1.10.5

### Technologies:
  * **Backbone**
  * **Marionette**
  * **jQuery**
  * **Coffeescript**
  * **Mustache**

### Overview:
  * **Coffeescript: ~500 files**
  * **Mustache: ~400 files**
  * **Modules: ~670**

### Webpack:
  * Initial Build Time: 36.3s / 36.16s / 35.88s ~ 36.11s
  * Incremental Build Time: 21.95s / 22.33s / 21.99s ~ 22.09s

**webpack config**: https://gist.github.com/maximerassi/e446f5260443b4b17a74

### Using exclude on loaders
Tried excluding ``node_modules`` and ``bower_components`` in each of the loaders.

```
{
    test: /\.mustache$/,
    loader: 'mustache'
    exclude: [path.join(rootPath, npmSrc), path.join(rootPath, bowerSrc)]
}
```

|                   | Before      | After                             | Improvement
|-------------------|-------------| ----------------------------------|------------
| Initial Build     | 36.11s      | 35.89s / 35.99s / 36.13s ~ 36.00s |  0.11s
| Incremental Build | 22.09s      | 22.27s / 21.89s / 22.16s ~ 22.11s | -0.02s

### LimitChunkCountPlugin
Although we would prefer an environment as close to production while developing, we are ok with this trade off.

|                   | Before    | After                               | Improvement
|-------------------|-----------| ------------------------------------|-------------
| Initial Build     | 37s       | 22.99s / 22.55s / 22.85s ~ 22.80s   | 14.2s
| Incremental Build | 17s       | 14.4s / 14.44s / 14.77s ~ 14.54s    | 2.46s

```
plugins: [
  new webpack.optimize.LimitChunkCountPlugin({
    maxChunks: 1
  })
],
```

### Adding main libraries to noParse
```
var deps = [
  'backbone/backbone-min.js',
  'backbone.babysitter/backbone.babysitter.min.js',
  'backbone.cocktail/Cocktail-0.5.10.min.js',
  'backbone.marionette/lib/backbone.marionette.min.js',
  'backbone.wreqr/backbone.wreqr.min.js',
  'underscore/underscore-min.js',
  'jquery/dist/jquery.min.js'
];

// Run through deps and extract the first part of the path,
// as that is what you use to require the actual node modules
// in your code. Then use the complete path to point to the correct
// file and make sure webpack does not try to parse it
deps.forEach(function (dep) {
  var depPath = path.resolve(npmSrc, dep);
  webpackConfig.resolve.alias[dep.split(path.sep)[0]] = depPath;
  webpackConfig.module.noParse.push(depPath);
});
// kudos to christianalfoni

```
|                   | Before    | After                               | Improvement
|-------------------|-----------| ------------------------------------|-------------
| Initial Build     | 36.11s    | 35.70s / 36.07s / 36.48 ~ 36.07s    | 0.04s
| Incremental Build | 22.09s    | 21.98s / 21.89s / 21.58s ~ 21.82s   | 0.27s

### CommonsChunkPlugin
I understood it only makes sense to use CommonsChunk with multiple entries app but we've also seen there could be potential incremental build improvements so gave it a shot anyways.

```
plugins: [
  new webpack.optimize.CommonsChunkPlugin('vendor', 'vendor.js', Infinity)
],
```

|                   | Before    | After                               | Improvement
|-------------------|-----------| ------------------------------------|-------------
| Initial Build     | 36.11s    | 35.88s / 35.80s / 35.69s ~ 35.79s   | 0.32s
| Incremental Build | 22.09s    | 21.74s / 22.02s / 21.98s ~ 21.94s   | 0.15s

### cache true
We are using webpack watch so this should be on by default.

```
var webpackConfig = {
    cache: true
    ...
```

### debug false

```
var webpackConfig = {
    debug: false
    ...
```

|                   | Before    | After                               | Improvement
|-------------------|-----------| ------------------------------------|-------------
| Initial Build     | 36.11s    | 36.52s / 37.86s / 35.91s ~ 36.43s   | -0.32s
| Incremental Build | 22.09s    | 22.21s / 22.35s / 22.72s ~ 22.43s   | -0.34s

### devtool to eval

```
var webpackConfig = {
    devtool: 'eval'
    ...
```

|                   | Before    | After                               | Improvement
|-------------------|-----------| ------------------------------------|-------------
| Initial Build     | 36.11s    | 37.17s / 35.96s / 36.4s ~ 36.51s    | -0.4s
| Incremental Build | 22.09s    | 22.41s / 22.12s / 22.95s ~ 22.49s   | -0.4s

### Hot Pack Dev Server

|                   | Before    | After                               | Improvement
|-------------------|-----------| ------------------------------------|-------------
| Initial Build     | 36.11s    | 34s / 34s / 34s ~ 34s               | 2.11s
| Incremental Build | 22.09s    | 23s / 27s / 21s ~ 23.67s            | -1.58s

### --profile --json
Looking at the chunks certainly feel like something might be wrong in the architecture
![2015-08-03_1-27-52](https://cloud.githubusercontent.com/assets/1452653/9066538/493769e4-3aa6-11e5-9d90-8c1dc8f44dca.jpg)
?

The results were compiled on Win 8.1 64 bit machine.

Any tips on optimizing this whether plugins or design changes would be greatly appreciated.

**EDIT:** We essentially have these requires in almost every file, could this be the bottleneck? We liked the idea of every single file defining its dependencies but if its going to start affecting performance we can live without it.
```
    _ = require 'g.underscore'
    commands = require 'g.bus.commands'
    events = require 'g.bus.events'
    reqres = require 'g.bus.reqres'
    Backbone = require 'g.backbone'
```

**Comment by sokra** [comment]
The chunks really look a bit weird. With the described structure your chunks should from a tree. So here is propably something wrong. You can click the chunk id for details. In these details `origins` propably give a good hint. Also follow the `reasons` of the modules. The chunks should not have so many parents.

Loader and parsing performance are catched by caching. So `noParse` will help to reduce initial building performance and won't affect incremental performance much.

You must use the `CommonsChunkPlugin` with your configration of `vendor` entry point. Elsewise libraries would be included multiple times. The `CommonsChunkPlugin` establish a parent-child relationship between `vendor` chunk and `initialize` chunk.

Interessting to see would be if any module has the `[non cachable]` flag and how many chunks have the `[rendered]` flag on an incremental build.

How many pages `page_x` and how many modules `module_x` do you have?

The chunk count should be `page count + module count + 2`.

**Comment by maximerassi** [comment]
Hi Tobia, thank you for your feedback. I started exploring it and as you spotted the chunks have a ton of parents. 

 * CommonsChunkPlugin. I just set it up but it did not seem to affect performance much.
``new webpack.optimize.CommonsChunkPlugin("vendor", "vendor.js")``

 *  I searched in the profiled json for ``"cacheable": false`` and did not find any matches, all seem to be set to ``"cacheable": true``

 * We are looking at 7 pages (x3 modes, create/show/edit) with around 14 modules (x3 modes create/show/edit)  (diagram was a bit oversimplified). Note, modules are not mutually exclusive to a page, they can and in some cases are re-used across pages. 

In fact I just started debugging from the entry point down and I just hit this dynamic require line that looks to be certainly one major bottleneck, it looks something like:

```
new: (type) ->
        require.ensure [], (require) ->
          slug = reqres.request "types:getSlugFromType", type
          dynamicApp = require './' + slug + '/' + slug + '_page'
          dynamicApp.new
```

From the URL it matches the /new/{type} route then loads appropriate page_x. Is this no good? I just looked more into it and saw the dynamic requires can be restricted to not search recursively, maybe that's the bottleneck? Too early to say I will be exploring more.

Thank you for your help and this great tool.

**Comment by maximerassi** [comment]
Hi @sokra our team ended up fixing the performance issues with code re-factoring and not so much webpack configuration. Essentially the solution was as you pointed out, reducing the number of parents for the modules (these modules were designed by nature to be extensively reused across pages and sometimes several times on the same page). This created way too many possibilities of re-use which slowed down re-compilation tremendously. With the new re-factoring we removed several require.ensures at the lowest levels (module level), we got it down to an initial build time of 15s and incremental build time of 1.2s, very good stuff! Thanks!

Here is how it now looks
![2015-08-09_21-42-09](https://cloud.githubusercontent.com/assets/1452653/9158190/1617dde4-3ee0-11e5-88fa-4f9b84e09568.jpg)
